### ISSUE TITLE
Make Events Cancelable

### ISSUE THREAD

**Issue by fionera** [Javacord/Javacord#383]
Currently when you have multiple MessageCreateEventListener they will all be called, even if you delete the Message in another Listener. It should be possible to cancel a event, so the other Listeners are not getting called.

**Comment by Saladoc** [comment]
This is problematic for two reasons:

1) Event cancellation implies that the cause for the event (a Message being created) is cancelled, which it is not at the moment, it's just a message waiting to be deleted.
2) The order in which event listeners are executed is not guaranteed, which might lead to inconsistent behavior depending on the order of listener execution.

I wonder what a use case for this might be.

**Comment by fionera** [comment]
Its just an idea. Another take on this would be to have a "isDeleted" function on the Entity

**Comment by Vampire** [comment]
@Saladoc I don't quite agree, actually I like the idea I think, just not the wording.

A better wording would probably be `consume()` and `isConsumed()`, so you can have listeners that obey the property and ones that ignore it, e. g. for logging.

This would be like it is e. g. for Swing events.

The use-case is e. g. you have a spam filter that runs first and consumes the event if it is spam.

Or you have one listener that consumes any event that is originated by yourself or a bot or a webhook, so that you don't have to check in the other listeners. And so on.

As the order actually is guaranteed. To quote the JavaDoc: `Adding a listener multiple times will only add it once and return the same listener manager on each invocation. The order of invocation is according to first addition.`.

So actually we only need a consumed property on the events base class, a setter and a getter and that's it. No other code needs to be changed as listeners should be able to ignore the property.

Alternatively the property instead of being a boolean could be some free Object, so that the event could be consumed for different aspects. Or maybe better, not having a consumed property, but a generic way to add information to an event for later listeners to process. So one listener can store that the event contains spam, another one could store it is not to be processed further, another one could store any other value and later listeners can use these information for the processing.

@fionera `isDeleted()` would not work as you expect. You can easily check whether a message is deleted in the meantime by calling `getLatestInstance​()` and checking whether you get a result in the future. But having called `delete()` on the message doesn't mean it is immediately deleted, it just enqueues a request to delete the message, that's why you get a future returned. It might even fail to do so which you only recognize if you actively handle exceptions.

**Comment by Saladoc** [comment]
@Vampire How is the execution order regarding to global vs server-attached vs channel-attached listeners? This would be relevant.

**Comment by Vampire** [comment]
Good point, I have to check in the generated class when I'm back from holidays or you look there yourself. It should always be the same defined order and that could be documented and a test added easily. I don't remember how I implemented it. Alternatively or additionally, the listeners could get a priority property, so that the user can define the order by defining a priority value and the defined order then is only for listeners with the same priority.

**Comment by Bastian** [comment]
@Saladoc
More "specific" listeners are always executed before less "specific".

**Comment by Vampire** [comment]
I think currently it is object attached first, globally attached last and within the object attached ordered alphabetically, not by specific-ness which sometimes is hard to do anyway. A channel is more specific than a server, but what is more specific, a message or a user?

**Comment by Bastian** [comment]
I like this idea of this issue but like Vampire I disagree with the naming.
I suggest `event.stopPropagation();` like it is common for Javascript events: https://developer.mozilla.org/en-US/docs/Web/API/Event/stopPropagation

But this only makes sense if we make it possible to modify the execution order of events (see https://github.com/Javacord/Javacord/issues/384).
I've also added an issue for adding custom information to events: https://github.com/Javacord/Javacord/issues/543
